File: dashboard/public/js/shared/pagination.js
window.TRKPagination = {
  createState: function (pageSize = 25) {
    return {
      page: 1,
      pageSize: pageSize,
      search: "",
      total: 0,
      hasNext: false,
      hasPrevious: false
    };
  },

  updateFromResponse: function (state, response) {
    state.page = response.page || state.page;
    state.pageSize = response.page_size || state.pageSize;
    state.total = response.total || 0;
    state.hasNext = !!response.has_next;
    state.hasPrevious = !!response.has_previous;
  },

  controlsHtml: function (idPrefix, state) {
    const totalPages = Math.max(1, Math.ceil((state.total || 0) / state.pageSize));

    return `
      <div class="dashboard-pagination" id="${idPrefix}Pagination">
        <button class="dashboard-btn dashboard-btn-light" id="${idPrefix}Prev" ${!state.hasPrevious ? "disabled" : ""}>
          Previous
        </button>

        <span class="dashboard-pagination-label">
          Page ${state.page} of ${totalPages}
        </span>

        <button class="dashboard-btn dashboard-btn-light" id="${idPrefix}Next" ${!state.hasNext ? "disabled" : ""}>
          Next
        </button>
      </div>
    `;
  },

  searchHtml: function (idPrefix, placeholder) {
    return `
      <div class="dashboard-table-search">
        <input
          type="search"
          id="${idPrefix}Search"
          class="dashboard-input"
          placeholder="${placeholder || "Search..."}"
        />
      </div>
    `;
  },

  bindControls: function (idPrefix, state, reloadFn) {
    const prevBtn = document.getElementById(`${idPrefix}Prev`);
    const nextBtn = document.getElementById(`${idPrefix}Next`);
    const searchInput = document.getElementById(`${idPrefix}Search`);

    if (prevBtn) {
      prevBtn.addEventListener("click", function () {
        if (!state.hasPrevious) return;
        state.page -= 1;
        reloadFn();
      });
    }

    if (nextBtn) {
      nextBtn.addEventListener("click", function () {
        if (!state.hasNext) return;
        state.page += 1;
        reloadFn();
      });
    }

    if (searchInput && !searchInput.dataset.bound) {
      searchInput.dataset.bound = "1";

      let searchTimer = null;

      searchInput.addEventListener("input", function () {
        clearTimeout(searchTimer);

        searchTimer = setTimeout(function () {
          state.search = searchInput.value || "";
          state.page = 1;
          reloadFn();
        }, 350);
      });
    }
  }
};

File: dashboard/api/shared/pagination.py
import math
import frappe


def get_page_args(default_page_size=25, max_page_size=100):
    try:
        page = int(frappe.form_dict.get("page") or 1)
    except Exception:
        page = 1

    try:
        page_size = int(frappe.form_dict.get("page_size") or default_page_size)
    except Exception:
        page_size = default_page_size

    page = max(page, 1)
    page_size = min(max(page_size, 1), max_page_size)

    return {
        "page": page,
        "page_size": page_size,
        "start": (page - 1) * page_size,
        "search": (frappe.form_dict.get("search") or "").strip(),
    }


def make_pagination(total, page, page_size):
    total_pages = max(1, math.ceil((total or 0) / page_size))

    return {
        "total": total or 0,
        "page": page,
        "page_size": page_size,
        "total_pages": total_pages,
        "has_previous": page > 1,
        "has_next": page < total_pages,
        "previous_page": page - 1 if page > 1 else 1,
        "next_page": page + 1 if page < total_pages else total_pages,
    }
